### PR TITLE
http2: migrates metadata_encoder_test.cc away from direct usage of nghttp2

### PR DESCRIPTION
**Pull request by birenroy**
Signed-off-by: Biren Roy <birenroy@google.com>

Part of addressing https://github.com/envoyproxy/envoy/issues/23596.

Commit Message: Migrates metadata_encoder_test.cc away from direct usage of nghttp2.
Additional Description:
Risk Level: low
Testing: ran the unit tests locally
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Comment by bencebeky** [comment]
FYI spell check failed:
```
./test/common/http/http2/metadata_encoder_test.cc:97:    // nghttp2_session_callbacks_del(callbacks);
```

Run `tools/spelling/check_spelling_pedantic.py` to check or fix it.

**Review by RyanTheOptimist — DISMISSED** [review]

**Comment by birenroy** [comment]
/assign @KBaichoo 
This one's ready, as soon as it has a deps review.

File: test/per_file_coverage.sh [hunk 1]
```diff
@@ -4,7 +4,7 @@
 # for existing directories with low coverage.
 declare -a KNOWN_LOW_COVERAGE=(
 "source/common:96.0"
-"source/common/api:82.5"
+"source/common/api:82.4"
```
Comment (RyanTheOptimist) [inline test/per_file_coverage.sh:7]: Is this change needed? I suspect it might be an artifact of a PR which landed yesterday and was reverted. Since your PR doesn't touch source/common/api, I don't *think* it need to adjust coverage. (Maybe this PR just needs a main merge?)

**Review by RyanTheOptimist — APPROVED** [review]

**Review by KBaichoo — APPROVED** [review]

**Review by bencebeky — APPROVED** [review]

**Review by diannahu — APPROVED** [review]
